clorm/orm/core.py

```diff
@@ -2360,115 +2357,223 @@ def __iter__(self):
+                out.append(space+l)
+        return "\n".join(out)
+
+    if not template:
+        return ''
+    lines = template.expandtabs(4).splitlines()
+    outlines = []
+    for line in lines:
+        start = line.find("{%")
+        if start == -1:
+            outlines.append(line)
+            continue
+        end = line.find("%}",start)
+        if end == -1:
+            raise ValueError("Bad template expansion in {line}")
+        kw = line[start+2:end]
+        text = add_spaces(start, kwargs[kw])
+        line = line[0:start] + text + line[end+2:]
+        outlines.append(line)
+    return "\n".join(outlines)
+
+PREDICATE_TEMPLATE = r"""
+def __init__(self,
+             {{%args_signature%}}
+             *, sign: bool=True, raw: AnySymbol=None):
+
+    self._hash = None
+    self._sign = bool(sign)
+
+    # If raw is supplied then we're using the Internal unification API so we
+    # don't do any value checking. So be warned that mis-using this API can
+    # result in unexcpected behaviour.
+    if raw is not None:
+        self._raw = raw
+        self._field_values = ({{%args%}})
+        return
+
+    {{%sign_check%}}
+    {{%check_no_defaults%}}
+
+    # Assign defaults for missing values and apply map tuple transform for complex values
+    {{%assign_defaults%}}
+    {{%check_complex%}}
+
+    self._field_values = ({{%args%}})
+
+    # Create the raw symbol
+    self._raw = _get_symbols().Function("{pdefn.name}",
+                                        ({{%args_raw%}}),
+                                        self._sign)
+
+@classmethod
+def _unify(cls: Predicate, raw: AnySymbol):
     try:
-        cls=type(self)
-        if (raw.name != cls.meta.name or
-            cls.meta.arity != len(raw.arguments) or
-            (cls.meta.sign is not None and cls.meta.sign != raw.positive)):
-            raise ValueError(("Failed to unify clingo.Symbol object {} with "
-                              "Predicate class {}").format(raw, cls.__name__))
-        self._sign = raw.positive
-        self._field_values = tuple( f.defn.cltopy(raw.arguments[f.index]) \
-                                    for f in self.meta )
-    except (TypeError,AttributeError,RuntimeError):
-        raise ValueError(("Failed to unify clingo.Symbol object {} with "
-                          "Predicate class {}").format(raw, cls.__name__))
-
-# Construct a Predicate via the field keywords
-def _predicate_init_by_keyword_values(self, **kwargs):
-    argnum=0
-    field_values = []
-    clingoargs = []
-    for f in self.meta:
-        if f.name in kwargs:
-            v= _preprocess_field_value(f.defn, kwargs[f.name])
-            argnum += 1
-        elif f.defn.has_default:
-            # Note: must be careful to get the default value only once in case
-            # it is a function with side-effects.
-            v = _preprocess_field_value(f.defn, f.defn.default)
+        raw_args = raw.arguments
+        if len(raw_args) != {pdefn.arity}:
```

Comment:
I haven't tested this experimentally, but all the sources I can find (eg. https://stackoverflow.com/questions/1115313/cost-of-len-function) say that the cost of len() for lists/tuples is O(1) because the length information is maintained explicitly within the list data structure. I think comparing two strings will always be O(n) in the worst-case since you have to compare each pair of letters.

Reply:
Ok, i just thought because in both case a call to some c-Function happens and they perhaps have a different execution time. But i think i'm trying to do micro-optimization here instead of searching for real bottlenecks...